src/docbuilder/metadata/lib.rs

```diff
@@ -0,0 +1,631 @@
+
+        // Let people opt-in to only having specific targets
+        let mut targets: HashSet<_> = self
+            .targets
+            .as_ref()
+            .map(|targets| targets.iter().map(String::as_str).collect())
+            .unwrap_or_else(|| DEFAULT_TARGETS.iter().copied().collect());
+
+        targets.remove(&default_target);
+        BuildTargets {
+            default_target,
+            other_targets: targets,
+        }
+    }
+
+    /// Return the arguments that should be passed to `cargo`.
+    ///
+    // TODO: maybe it shouldn't?
+    /// This will always include `doc --lib --no-deps`.
+    /// This will never include `--target`.
+    ///
+    /// Note that this does not necessarily reproduce the HTML _output_ of docs.rs exactly.
+    /// For example, the links may point somewhere different than they would on docs.rs.
+    /// However, rustdoc will see exactly the same code as it would on docs.rs, even counting `cfg`s.
+    pub fn cargo_args(&self) -> Vec<String> {
+        let mut cargo_args: Vec<String> = vec!["doc".into(), "--lib".into(), "--no-deps".into()];
+
+        if let Some(features) = &self.features {
+            cargo_args.push("--features".into());
+            cargo_args.push(features.join(" "));
+        }
+
+        if self.all_features {
+            cargo_args.push("--all-features".into());
+        }
+
+        if self.no_default_features {
+            cargo_args.push("--no-default-features".into());
+        }
+
+        cargo_args
+    }
+
+    /// Return the environment variables that should be set when building this crate.
+    pub fn environment_variables(&self) -> HashMap<&'static str, String> {
+        let joined = |v: &Option<Vec<_>>| v.as_ref()
+            .map(|args| args.join(" "))
+            .unwrap_or_default();
+
+        let mut map = HashMap::new();
+        map.insert("RUSTFLAGS", joined(&self.rustc_args));
+        map.insert("RUSTDOCFLAGS", joined(&self.rustdoc_args));
+        // For docs.rs detection from build scripts:
+        // https://github.com/rust-lang/docs.rs/issues/147
+        map.insert("DOCS_RS", "1".into());
+
+        map
+    }
+}
+
+impl std::str::FromStr for Metadata {
+    type Err = toml::de::Error;
+
+    /// Parse the given manifest as TOML.
+    fn from_str(manifest: &str) -> Result<Metadata, Self::Err> {
+        let mut metadata = Metadata::default();
+
+        let manifest = manifest.parse::<Value>()?;
+
+        fn fetch_manifest_tables<'a>(manifest: &'a Value) -> Option<&'a Map<String, Value>> {
+            manifest
+                .get("package")?
+                .as_table()?
+                .get("metadata")?
+                .as_table()?
+                .get("docs")?
+                .as_table()?
+                .get("rs")?
+                .as_table()
+        }
+
+        if let Some(table) = fetch_manifest_tables(&manifest) {
+            // TODO: all this `to_owned` is inefficient, this should use explicit matches instead.
```

Comment:
Could this block just be [`table.try_into()?`](https://docs.rs/toml/0.5.6/toml/value/enum.Value.html#method.try_into) with `derive(Deserialize)`?

Reply:
I'm not sure what you mean here. This is a `Value` we already have, not a string, there's no deserializers in `toml` I see that take a `Value`.

Reply:
Oh I missed the link - good idea, I'll try that

Reply:
Done!